tests/unittests/test_exception_handling.py

```diff
@@ -0,0 +1,331 @@
+        xero_client.access_token = "123"
+        xero_client.tenant_id = "123"
+
+        try:
+            xero_client.filter(tap_stream_id)
+        except client_.XeroInternalError as e:
+            expected_error_message = "HTTP-error-code: 500, Error: An unhandled error with the Xero API. Contact the Xero API team if problems persist."
+
+            # Verifying the message formed for the custom exception
+            self.assertEquals(str(e), expected_error_message)
+            pass
+
+
+    @mock.patch('requests.Request', side_effect=mocked_notimplemented_501_error)
+    def test_notimplemented_501_error(self, mocked_session, mocked_notimplemented_501_error):
+        config = {}
+        tap_stream_id = "contacts"
+
+        xero_client = client_.XeroClient(config)
+        xero_client.access_token = "123"
+        xero_client.tenant_id = "123"
+
+        try:
+            xero_client.filter(tap_stream_id)
+        except client_.XeroNotImplementedError as e:
+            expected_error_message = "HTTP-error-code: 501, Error: The method you have called has not been implemented."
+
+            # Verifying the message formed for the custom exception
+            self.assertEquals(str(e), expected_error_message)
+            pass
+
+
+    @mock.patch('requests.Request', side_effect=mocked_failed_429_request)
+    def test_too_many_requests_429_error(self, mocked_session, mocked_failed_429_request):
+        config = {}
+        tap_stream_id = "contacts"
+
+        xero_client = client_.XeroClient(config)
+        xero_client.access_token = "123"
+        xero_client.tenant_id = "123"
+
+        try:
+            # Verifying if the custom exception 'XeroTooManyError' is raised on receiving status code 429
+            filter_func_exec = xero_client.filter(tap_stream_id)
+        except client_.XeroTooManyError as e:
+            expected_error_message = "HTTP-error-code: 429, Error: The API rate limit for your organisation/application pairing has been exceeded. Please retry after 10 seconds"
+            
+            # Verifying the message formed for the custom exception
+            self.assertEquals(str(e), expected_error_message)
+            pass
+
+
+    @mock.patch('requests.Request', side_effect=mocked_failed_429_request)
+    def test_too_many_requests_429_backoff_behavior(self, mocked_session, mocked_failed_429_request):
+        config = {}
+        tap_stream_id = "contacts"
+
+        xero_client = client_.XeroClient(config)
+        xero_client.access_token = "123"
+        xero_client.tenant_id = "123"
+        try:
+            filter_func_exec = xero_client.filter(tap_stream_id)
+        except (requests.HTTPError, client_.XeroTooManyError) as e:
+            pass
+
+        self.assertEqual(mocked_failed_429_request.call_count, 3)
+        self.assertEqual(mocked_session.call_count, 3)
+
+
+
+@mock.patch('requests.Session.send', side_effect=mocked_session)
+class TestCheckPlatformAccessBehavior(unittest.TestCase):
+
```

Comment:
TDL-594 (Verifying the authentication and authorization behavior during discovery mode)